Fix CI and unstable transport

aarani opened the pull request:
Recently CI has been extremely buggy and unreliable, during extensive testing TCP Transport is identified as the culprit.
Errors like 
> System.InvalidOperationException : Couldn't read the packet length (was 0, expected 4)

that would happen because of force disconnections right after the first encrypted write.
For this reason and because the transport layer was long overdue for a rewrite we changed the transport to an asynchronous websocket transport + aes-ctr obfuscation

aarani committed:
Core: port AES-256-CTR from BoringSSL

omerjacobi commented:
@aarani do you know why the CI for windows still fails?

aarani commented:
> @aarani do you know why the CI for windows still fails?

nblockchain's CI sercrets needs an update, for checking out the actual CI result check the results on my fork.

studasd commented:
@aarani
New transport will work through proxy?

aarani commented:
yeah, It's not implemented but with a simple code change you can use proxies.

aarani committed:
Tests: load NumberToSendMessage in init process

738a849 moves loading of "NumberToSendMessage" property
to SendMessage test even though there are 3 other tests
that use that variable thereby breaking all 3 tests
because of NumberToSendMessage being null.

This commit moves the load process back to init phase so
all 4 tests can use NumberToSendMessage.

knocte commented on src/TgSharp.Core/Auth/Authenticator.cs:
```diff
@@ -1,45 +1,61 @@
-            await sender.Send(step1.ToBytes(), token).ConfigureAwait(false);
-            var step1Response = step1.FromBytes(await sender.Receive(token)
-                .ConfigureAwait(false));
+            //race condition ? move randomization in consturctor
```
please add more details about this potential race condition

knocte commented on src/TgSharp.Core/Auth/Authenticator.cs:
```diff
@@ -1,45 +1,61 @@
+        }
 
-            var sender = new MtProtoPlainSender(transport);
+        public Task<Step3_Response> DoAuthentication()
```
if this returns Task, shouldn't it be marked as async?

aarani replied:
no you don't have to when you already have a Task, making the function async and awaiting a Task just adds more overhead.

knocte replied:
what do you mean more overhead? the way this method is consumed is actually via an await so it should be marked as async

aarani replied:
https://stackoverflow.com/questions/38017016/async-task-then-await-task-vs-task-then-return-task

knocte commented on src/TgSharp.Core/Auth/Authenticator.cs:
```diff
@@ -1,45 +1,61 @@
+        }
 
-            return step3Response;
+        private void Sender_OnStep3ResponseReceived(Step3_CompleteDHExchange step3, byte[] response)
```
please don't use underscore in any names

aarani replied:
underscore in the class is unchanged from the parent, underscore in the event function name is standard naming in c#

knocte commented on src/TgSharp.Core/MTProto/Crypto/AesCtr.cs:
```diff
@@ -0,0 +1,96 @@
+        // of the IV.  This implementation takes NO responsibility for checking that
+        // the counter doesn't overflow into the rest of the IV when incremented.
+
+        public static void Ctr128Encrypt(byte[] input, byte[] key, ref byte[] ivec, ref byte[] ecount_buf, ref int num, byte[] output)
```
better name: eCountBuffer, not ecount_buf, or even also unabbreviate 'e'

aarani replied:
fixed

knocte commented on src/TgSharp.Core/MTProto/Crypto/AesCtr.cs:
```diff
@@ -0,0 +1,96 @@
+
+        private static byte[] EncryptBlock(byte[] toEncrypt, byte[] key)
+        {
+            using (RijndaelManaged aes = new RijndaelManaged())
```
nit: use var for less redundancy

aarani replied:
fixed

knocte commented on src/TgSharp.Core/Network/MtProtoPlainSender.cs:
```diff
@@ -7,21 +7,38 @@ namespace TgSharp.Core.Network
 {
     public class MtProtoPlainSender
     {
+        public delegate void ResponseHandler(byte[] response);
```
delegates are old school, just use Funcs or Actions

aarani replied:
should be fixed

knocte commented on src/TgSharp.Core/Network/MtProtoPlainSender.cs:
```diff
@@ -7,21 +7,38 @@ namespace TgSharp.Core.Network
         {
-            token.ThrowIfCancellationRequested();
+            using (MemoryStream memoryStream = new MemoryStream(message.Body))
+            using (BinaryReader binaryReader = new BinaryReader(memoryStream))
```
nit: use var in the 2 lines above

aarani replied:
should be fixed

knocte commented on src/TgSharp.Core/Network/MtProtoSender.cs:
```diff
@@ -15,41 +15,88 @@
+            using (var inputReader = new BinaryReader(inputStream))
+            {
+                if (inputReader.BaseStream.Length < 8)
+                    throw new InvalidOperationException($"Can't decode packet");
```
no need for $

aarani replied:
fixed

knocte commented on src/TgSharp.Core/Network/MtProtoSender.cs:
```diff
@@ -272,86 +293,74 @@ private bool HandleGzipPacked(ulong messageId, int sequence, BinaryReader messag
+                catch (Exception ex)
                 {
-                    throw new InvalidOperationException(errorMessage);
+                    SetRequsetException(requestId, ex);
```
typo here

aarani replied:
fixed

knocte commented on src/TgSharp.Core/Network/MtProtoSender.cs:
```diff
@@ -547,10 +538,50 @@ private bool HandleContainer(ulong messageId, int sequence, BinaryReader message
         }
+        #endregion
+
+        #region Helper functions
```
regions are a smell, create a new class

aarani replied:
I'll just remove the regions just wanted to be able to collapse multiple functions at once while I was working on them

knocte commented on src/TgSharp.Core/Network/Requests/AckRequest.cs:
```diff
@@ -6,10 +6,12 @@
     {
         private readonly List<ulong> msgs;
 
+        public override int Constructor => 0x62d6b459;
```
what is this magic number?

aarani replied:
Each TLObject-derived class has a constructor it's part of the Telegram Protocol. check any class of TLSharp.TL for more example.

knocte commented on src/TgSharp.Core/Network/TgMessage.cs:
```diff
@@ -0,0 +1,46 @@
+using System.Linq;
+using TgSharp.Core.MTProto.Crypto;
+
+namespace TgSharp.Core.Network
```
drop the Tg prefix, we already have this technology called namespaces

knocte commented on src/TgSharp.Core/Network/TgMessage.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace TgSharp.Core.Network
+{
+    public class TgMessage
```
drop the Tg prefix, we already have this technology called namespaces

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,173 @@
+        private void ReconnectionHappened(ReconnectionInfo obj)
+        {
+            if (client.IsRunning)
+                InitializeConnection().Wait();
```
if InitializeConnection is async, thten make ReconnectionHappened async too

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,173 @@
+
+        public async Task ConnectAsync(int retryCount = 5)
+        {
+
```
nit: drop this EOL

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,173 @@
+
+        public void Send(byte[] packet)
+        {
+
```
nit: drop this EOL

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,173 @@
+
+        public void Dispose()
+        {
+            client.Dispose();
```
assign client to null afterwards?

knocte commented on src/TgSharp.Core/TgSharp.Core.csproj:
```diff
@@ -9,7 +9,7 @@
     <RootNamespace>TgSharp.Core</RootNamespace>
     <AssemblyName>TgSharp.Core</AssemblyName>
-    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
```
why? add the reason in commit message

knocte commented on src/TgSharp.TL/TL/Account/TLContentSettings.cs:
```diff
@@ -26,7 +26,10 @@ public override int Constructor
-            // do nothing
+            Flags = 0;
+            Flags = SensitiveEnabled ? (Flags | 1) : (Flags & ~1);
+            Flags = SensitiveCanChange ? (Flags | 2) : (Flags & ~2);
```
can separate the flags fix in a different commit pls?

aarani replied:
it's separated now

knocte replied:
so does the PR work if you separate this commit to a new PR?

aarani replied:
yeah removed it from the PR

knocte commented on src/TgSharp.Tests/packages.config:
```diff
@@ -1,3 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="System.Reactive" version="4.3.2" targetFramework="net472" />
```
what's the dep that needs System.Reactive?

aarani replied:
Websocket-client

knocte commented on src/TgSharp.Tests/TgSharp.Tests.csproj:
```diff
@@ -48,7 +69,7 @@
     </ProjectReference>
     <ProjectReference Include="..\TgSharp.TL\TgSharp.TL.csproj">
-      <Project>{D6144517-91D2-4880-86DF-E9FF5D7F383A}</Project>
+      <Project>{d6144517-91d2-4880-86df-e9ff5d7f383a}</Project>
```
revert this, your IDE is crazy

knocte commented on src/TgSharp.Tests/TgSharpTests.cs:
```diff
@@ -243,7 +244,7 @@ public virtual async Task DownloadFileFromContactTest()
 
             var inputPeer = new TLInputPeerUser() { UserId = user.Id };
-            var res = await client.SendRequestAsync<TLMessagesSlice>(new TLRequestGetHistory() { Peer = inputPeer });
+            var res = (TLMessagesSlice)await client.SendRequestAsync(new TLRequestGetHistory() { Peer = inputPeer });
```
why make this worse? why not generic?

aarani replied:
It's a bit complex to explain, but the gist of it is that SendRequestAsync didn't know what was the response type of the TLRequestGetHistory before so it had to get the response in generic, now it knows BUT the response type of TLRequestGetHistory is TLAbsMessages (an abstraction) which then should be casted accordingly

aarani commented:
> if it's only to be able to use .NETStandard, you can use 4.7.1

2 The versions listed here represent the rules that NuGet uses to determine whether a given .NET Standard library is applicable. While NuGet considers .NET Framework 4.6.1 as supporting .NET Standard 1.5 through 2.0, there are several issues with consuming .NET Standard libraries that were built for those versions from .NET Framework 4.6.1 projects. For .NET Framework projects that need to use such libraries, we recommend that you upgrade the project to target .NET Framework 4.7.2 or higher.

[Source](https://docs.microsoft.com/en-us/dotnet/standard/net-standard)

knocte commented:
> [Source](https://docs.microsoft.com/en-us/dotnet/standard/net-standard)

I was suggesting 4.7.1 because it's what we use in geewallet and works fine. Anyway this is a good source, please add the URL to the commit message.

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,178 @@
+                    await InitializeClient();
+                    break;
+                }
+                catch (Exception ex)
```
what kind of exception can happen here? let's put the specific type, not generic Exception base class

aarani replied:
not sure but the point is that in case of any error I want to retry there are a lot of socket related exceptions

knocte replied:
can you gist 2 exceptions of this, and make retryCount default to 0

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,178 @@
+
+        private async Task InitializeClient()
+        {
+            client = new WebsocketClient(new Uri($"ws://{address}:{port}/apiws"));
```
can you rather assign this in the constructor?

aarani replied:
no, there is a bug in the library we use (tbh I don't know if it's a bug or expected behavior) that when you call .StartOrFail() it sets a flag as true and throws in case of an error but when you want to retry it because that flag is already true it doesn't even try, now that I think about it I might need to dispose it before replacing it

aarani replied:
added a dispose call before it

knocte replied:
I only said assign, I didnt mean you to call .StartOrFail() in the constructor

aarani replied:
I call InitializeClient in the retry loop for connecting, If I assign the client in constructor instead of InitializeClient the bug that I just said will happen

aarani replied:
so I have to reassign it to a new one every time

aarani replied:
which should be inside the loop and that's the best place for it

knocte commented on src/TgSharp.Core/Network/WSTransport.cs:
```diff
@@ -0,0 +1,181 @@
+            await client.StartOrFail();
+        }
+
+        private async Task ReconnectionHappened(ReconnectionInfo obj)
```
this is a bad name for this method; the name seems to suggest that it returns a bool without doing anything, but it actually does something, it reconnects! let's rename this method

knocte commented on src/TgSharp.Core/Session.cs:
```diff
@@ -39,6 +40,10 @@ internal static Session TryLoadOrCreateNew (ISessionStore store, string sessionU
+            session.Sequence = 0;
+            byte[] randomSessionId = new byte[8];
+            RandomNumberGenerator.Create().GetNonZeroBytes(randomSessionId);
+            session.Id = BitConverter.ToUInt64(randomSessionId, 0);
```
if this commit is supposed to fix the same problem that the CI define fixed, then the removal of CI define should be in the same commit as this

aarani commented:
Based on @Laiteux's testing, this commit might cause issues in .NET/.NET Core. These platforms are unsupported by us cause TgSharp is a .NET Framework library but I think it makes sense to fix it before merge

knocte commented:
> Based on @Laiteux's testing, this commit might cause issues in .NET/.NET Core. These platforms are unsupported by us cause TgSharp is a .NET Framework library but I think it makes sense to fix it before merge

We can fix it later when we migrate to .NETCore/.NET5.

Laiteux commented:
That means I won't be able to keep using TgSharp anymore, since all my projects are .NET 5 or at least .NET Core 3.1. Honestly, who uses .NET Framework anymore...

Or then a second option would be to keep the TcpTransport and let the user decide what he wanna use, but that would introduce more complexity than anything else.

knocte commented:
> I won't be able to keep using TgSharp anymore

I said we will migrate, but no need to do it in this PR. The smallest this PR is the better, and the earlier we can start merging other PRs. We don't want this PR to become a monster.

aarani commented:
> > Based on @Laiteux's testing, this commit might cause issues in .NET/.NET Core. These platforms are unsupported by us cause TgSharp is a .NET Framework library but I think it makes sense to fix it before merge
> 
> We can fix it later when we migrate to .NETCore/.NET5.

I agree! it's not our bug either, so we either have to try to fix our dependency and PR it or just use another websocket library which we can figure out after this gets merged

aarani commented:
> That means I won't be able to keep using TgSharp anymore, since all my projects are .NET 5 or at least .NET Core 3.1. Honestly, who uses .NET Framework anymore...
> 
> Or then a second option would be to keep the TcpTransport and let the user decide what he wanna use, but that would introduce more complexity than anything else.

It's just for this PR not permanent.

knocte commented on src/TgSharp.Core/Network/MtProtoPlainSender.cs:
```diff
@@ -7,21 +7,37 @@ namespace TgSharp.Core.Network
+        public MtProtoPlainSender(WSTransport transport)
         {
             this.transport = transport;
+            transport.OnUnencryptedMessage += Transport_OnUnencryptedMessage;
```
do we need to unsuscribe this at dispose/reconnection time? otherwise there's a leak

knocte replied:
ping ^

aarani committed:
Core: remove seqNo and sessionId from session file

As long as we create a new sessionId for every connection,
(by randomizing the sessionId) seqNo can start from 0
and we don't have to save it. This prevents errors
like code=32 that are caused by wrong seqNo from happening.

This commit also removes the previous workaround that was
used on CI to prevent these errors cause they are no longer
necessary.

Laiteux commented:
This will fix the retry logic:
```
@@ -136,19 +128,20 @@ namespace TgSharp.Core.Network
             await client.SendInstant(finalInit);
         }

-        public async Task ConnectAsync(int retryCount = 5)
+        public async Task ConnectAsync(int retryCount = 0)
         {
-            while (retryCount-- != 0)
+            while (retryCount-- != -1)
             {
                 try
                 {
                     await InitializeClient();
                     break;
                 }
-                catch (Exception ex)
+                catch
                 {
-                    if (retryCount == 1)
-                        throw ex;
+                    if (retryCount == -1)
+                        throw;
+
                     await Task.Delay(1000);
                 }
             }
```
I don't have access to this repo therefore I can't do the changes myself for now.

knocte commented:
CI is broken now in your branch.

aarani commented:
> CI is broken now in your branch.

because my session is now broken

knocte commented:
> because my session is now broken

Let's fix that before you change the retryCount.

omerjacobi commented:
any update on this PR?

aarani committed:
Core,Tests: Transport layer rewrite [2/2]

  - Upgrade .NET to 4.7.2 to use .NET Standard dependency
    Reason:
      https://docs.microsoft.com/en-us/dotnet/standard/net-standard
  - Uses obfuscated WebSocket transport instead of TCP
  - Adds asynchronous response handling

aarani committed:
Gen,TL: Transport layer rewrite [1/2]

 - add response type as generic

knocte commented on src/TgSharp.Core/Network/Message.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace TgSharp.Core.Network
+{
+    public class Message
```
@aarani nit, should this be internal too?

knocte commented on src/TgSharp.Core/Network/MtProtoSender.cs:
```diff
@@ -388,82 +406,58 @@ private bool HandleMsgDetailedInfo(ulong messageId, int sequence, BinaryReader m
-
-            uint code = messageReader.ReadUInt32();
-            ulong badMsgId = messageReader.ReadUInt64();
+            _ = messageReader.ReadUInt32();
```
in HandlePong, you have put a comment explaining why this is ignored (_), can you put a comment here too?

aarani replied:
all those comments are mostly old I just ignored some of the unused variables

knocte replied:
maybe it's easy to figure out from looking at the spec again for a minute?

knocte commented on src/TgSharp.Core/app.config:
```diff
@@ -0,0 +1,11 @@
+    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+      <dependentAssembly>
+        <assemblyIdentity name="System.Threading.Tasks.Extensions" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-4.2.0.1" newVersion="4.2.0.1" />
```
do all TgSharp consumers need this bindingRedirect now?

aarani replied:
I don't know to be honest, I don't think so though

knocte replied:
if this is the case, this is a big annoyance... I wonder if we can fix it

knocte commented:
Last but not least, I'm having trouble finding the MS docs where it recommends to use underscores in method names.

aarani commented:
> Last but not least, I'm having trouble finding the MS docs where it recommends to use underscores in method names.

which methods are you talking about ?

aarani commented:
> Last but not least, I'm having trouble finding the MS docs where it recommends to use underscores in method names.

I can't find the original documentation for it, but event handler name Object_EventName is a fairly old convention.

https://docs.microsoft.com/en-us/dotnet/desktop/winforms/event-handlers-overview-windows-forms?view=netframeworkdesktop-4.8

also VS creates event handlers with this same exact naming scheme

omerjacobi commented:
any update?

Laiteux commented:
> any update?

Nope, and there will never be. Because knocte always had a urge to control every single misplaced comma or underscore despite wanting to take the lead on **open-source** projects. Hopefully one day he understands you can't have a desire to lead people **and** a compulsive behavior to "optimize" everyone's code. I would have loved too as well, sadly you'll never see this PR being merged, mainly for that reason. Everyone eventually got tired and that's it, that's why this project is dead. Props to @aarani for his crazy work on everything else and from the very beginning, without that dude I probably wouldn't even be earning a living the way I'm doing it today.
